# Homework 4 Part 2
## Daniel O'Brien

### Data importing and preprocessing

In [1]:
import numpy as np
import pandas as pd

In [2]:
datMat = pd.read_csv('segmentation_data.txt', header = None)
datMat

0      1   2         3    4         5         6         7         8   \
0     110.0  189.0   9  0.000000  0.0  1.000000  0.666667  1.222222  1.186342   
1      86.0  187.0   9  0.000000  0.0  1.111111  0.720082  1.444444  0.750309   
2     225.0  244.0   9  0.000000  0.0  3.388889  2.195113  3.000000  1.520234   
3      47.0  232.0   9  0.000000  0.0  1.277778  1.254621  1.000000  0.894427   
4      97.0  186.0   9  0.000000  0.0  1.166667  0.691215  1.166667  1.005540   
...     ...    ...  ..       ...  ...       ...       ...       ...       ...   
2095   32.0  158.0   9  0.000000  0.0  0.944445  0.862963  0.833333  0.611111   
2096    8.0  162.0   9  0.111111  0.0  1.611111  2.062962  0.333333  0.133333   
2097  128.0  161.0   9  0.000000  0.0  0.555555  0.251852  0.777778  0.162963   
2098  150.0  158.0   9  0.000000  0.0  2.166667  1.633334  1.388889  0.418518   
2099  124.0  162.0   9  0.111111  0.0  1.388889  1.129630  2.000000  0.888889   

             9          10         11         12        13         14  \
0     12.925926  10.888889   9.222222  18.666668 -6.111111 -11.111111   
1     13.740741  11.666667  10.333334  19.222221 -6.222222 -10.222222   
2     12.259259  10.333334   9.333334  17.111110 -5.777778  -8.777778   
3     12.703704  11.000000   9.000000  18.111110 -5.111111 -11.111111   
4     15.592592  13.888889  11.777778  21.111110 -5.111111 -11.444445   
...         ...        ...        ...        ...       ...        ...   
2095   7.962963   6.333334  11.888889   5.666666 -4.888889  11.777778   
2096   8.370370   6.666666  12.000000   6.444445 -5.111111  10.888889   
2097   7.148148   5.555555  10.888889   5.000000 -4.777778  11.222222   
2098   8.444445   7.000000  12.222222   6.111111 -4.333334  11.333333   
2099  10.037037   8.000000  14.555555   7.555555 -6.111111  13.555555   

             15         16        17        18  
0     17.222221  18.666668  0.508139  1.910864  
1     16.444445  19.222221  0.463329  1.941465  
2     14.555555  17.111110  0.480149  1.987902  
3     16.222221  18.111110  0.500966  1.875362  
4     16.555555  21.111110  0.442661  1.863654  
...         ...        ...       ...       ...  
2095  -6.888889  11.888889  0.520578 -1.982834  
2096  -5.777778  12.000000  0.484805 -2.044946  
2097  -6.444445  10.888889  0.540918 -1.996307  
2098  -7.000000  12.222222  0.503086 -1.943449  
2099  -7.444445  14.555555  0.479931 -2.029312  

[2100 rows x 19 columns]

In [3]:
Names = pd.read_csv('segmentation_names.txt', header=None)
Names

0
0    REGION-CENTROID-COL
1    REGION-CENTROID-ROW
2     REGION-PIXEL-COUNT
3   SHORT-LINE-DENSITY-5
4   SHORT-LINE-DENSITY-2
5             VEDGE-MEAN
6               VEDGE-SD
7             HEDGE-MEAN
8               HEDGE-SD
9         INTENSITY-MEAN
10           RAWRED-MEAN
11          RAWBLUE-MEAN
12         RAWGREEN-MEAN
13            EXRED-MEAN
14           EXBLUE-MEAN
15          EXGREEN-MEAN
16            VALUE-MEAN
17       SATURATION-MEAN
18              HUE-MEAN

In [4]:
classes= pd.read_csv('segmentation_classes.txt', delimiter = '\t', header=None)
classes

0  1
0      GRASS  0
1      GRASS  0
2      GRASS  0
3      GRASS  0
4      GRASS  0
...      ... ..
2095  CEMENT  3
2096  CEMENT  3
2097  CEMENT  3
2098  CEMENT  3
2099  CEMENT  3

[2100 rows x 2 columns]

In [5]:
classes1 = classes.drop(columns=[0])
classes1

1
0     0
1     0
2     0
3     0
4     0
...  ..
2095  3
2096  3
2097  3
2098  3
2099  3

[2100 rows x 1 columns]

In [6]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import completeness_score, homogeneity_score


In [7]:
scaler = MinMaxScaler()

In [8]:
datMat = scaler.fit_transform(datMat)


In [9]:
datMat.shape

(2100, 19)

## Clustering before PCA

In [10]:
from sklearn.cluster import KMeans

In [11]:
train_kmeans = KMeans(n_clusters=7, max_iter=500, verbose=0, n_init=5) # initialization
train_kmeans.fit(datMat)

KMeans(max_iter=500, n_clusters=7, n_init=5)

In [12]:
cluster = train_kmeans.predict(datMat)

In [13]:
cluster

array([2, 2, 2, ..., 3, 3, 4], dtype=int32)

In [14]:
clusters = pd.DataFrame(cluster, columns=['Cluster'])
clusters

Cluster
0           2
1           2
2           2
3           2
4           2
...       ...
2095        6
2096        4
2097        3
2098        3
2099        4

[2100 rows x 1 columns]

In [15]:
pd.options.display.float_format='{:,.2f}'.format

centroids = pd.DataFrame(train_kmeans.cluster_centers_, columns=Names)
centroids

(REGION-CENTROID-COL,)  (REGION-CENTROID-ROW,)  (REGION-PIXEL-COUNT,)  \
0                    0.54                    0.15                   0.00   
1                    0.75                    0.53                   0.00   
2                    0.51                    0.81                   0.00   
3                    0.77                    0.43                   0.00   
4                    0.25                    0.39                   0.00   
5                    0.30                    0.53                   0.00   
6                    0.25                    0.46                   0.00   

   (SHORT-LINE-DENSITY-5,)  (SHORT-LINE-DENSITY-2,)  (VEDGE-MEAN,)  \
0                     0.03                     0.00           0.03   
1                     0.04                     0.04           0.11   
2                     0.08                     0.01           0.05   
3                     0.01                     0.02           0.04   
4                     0.08                     0.02           0.08   
5                     0.05                     0.05           0.10   
6                     0.03                     0.01           0.04   

   (VEDGE-SD,)  (HEDGE-MEAN,)  (HEDGE-SD,)  (INTENSITY-MEAN,)  (RAWRED-MEAN,)  \
0         0.00           0.03         0.00               0.82            0.78   
1         0.02           0.11         0.02               0.30            0.28   
2         0.00           0.05         0.00               0.11            0.09   
3         0.00           0.02         0.00               0.04            0.03   
4         0.00           0.06         0.01               0.15            0.14   
5         0.01           0.08         0.01               0.40            0.37   
6         0.00           0.03         0.00               0.03            0.02   

   (RAWBLUE-MEAN,)  (RAWGREEN-MEAN,)  (EXRED-MEAN,)  (EXBLUE-MEAN,)  \
0             0.89              0.79           0.27            0.67   
1             0.35              0.26           0.59            0.45   
2             0.09              0.14           0.68            0.08   
3             0.06              0.03           0.78            0.22   
4             0.18              0.12           0.72            0.34   
5             0.47              0.35           0.50            0.57   
6             0.04              0.02           0.77            0.22   

   (EXGREEN-MEAN,)  (VALUE-MEAN,)  (SATURATION-MEAN,)  (HUE-MEAN,)  
0             0.29           0.89                0.21         0.13  
1             0.31           0.35                0.30         0.16  
2             0.82           0.13                0.41         0.89  
3             0.49           0.06                0.54         0.24  
4             0.35           0.18                0.41         0.20  
5             0.21           0.47                0.30         0.16  
6             0.51           0.04                0.80         0.18

In [16]:
def cluster_sizes(clusters):
    #clusters is an array of cluster labels for each instance in the data
    
    size = {}
    cluster_labels = np.unique(clusters)
    n_clusters = cluster_labels.shape[0]

    for c in cluster_labels:
        size[c] = len(datMat[clusters == c])
    return size

In [17]:
size = cluster_sizes(cluster)

for c in size.keys():
    print("Size of Cluster", c, "= ", size[c])

Size of Cluster 0 =  300
Size of Cluster 1 =  332
Size of Cluster 2 =  297
Size of Cluster 3 =  309
Size of Cluster 4 =  336
Size of Cluster 5 =  236
Size of Cluster 6 =  290


In [18]:
classNP = classes1.values.flatten()
classNP.shape

(2100,)

In [19]:
print(completeness_score(classNP,cluster))

0.6116744999910891


In [20]:
print(homogeneity_score(classNP,cluster))

0.6099656393147241


## Completeness and Homogeneity Results
#### Before running PCA, the completeness score returned a value of 0.612 and the homogeneity returned a value of 0.610. The completeness score shows that rows belonging to the same class were placed in multiple clusters. And the homogeneity score of 0.610 shows that there are different classes belonging to the same clusters.

# PCA

In [21]:
from sklearn import decomposition

In [22]:
pca = decomposition.PCA()

In [23]:
dTrans = pca.fit(datMat).transform(datMat)

In [24]:
#Print the Projected Data
np.set_printoptions(precision=2,suppress=True)
print(dTrans)

[[-0.69  0.53  0.25 ... -0.    0.    0.  ]
 [-0.67  0.51  0.34 ...  0.   -0.    0.  ]
 [-0.71  0.77 -0.16 ...  0.    0.    0.  ]
 ...
 [-0.51 -0.13 -0.08 ... -0.    0.    0.  ]
 [-0.48 -0.09 -0.16 ... -0.    0.    0.  ]
 [-0.44 -0.11 -0.05 ... -0.    0.    0.  ]]


In [25]:
print(pca.explained_variance_ratio_)

[0.61 0.13 0.1  0.05 0.04 0.02 0.02 0.02 0.01 0.01 0.   0.   0.   0.
 0.   0.   0.   0.   0.  ]


In [26]:
## Run PCA again with keeping on 1 component - Training Data
pca1 = decomposition.PCA(n_components = 6)
dTrans1 = pca1.fit(datMat).transform(datMat)

In [27]:
np.set_printoptions(precision=2,suppress=True)
print(dTrans1)

[[-0.69  0.53  0.25 -0.2  -0.08  0.05]
 [-0.67  0.51  0.34 -0.17 -0.04  0.06]
 [-0.71  0.77 -0.16 -0.01 -0.17  0.04]
 ...
 [-0.51 -0.13 -0.08  0.1  -0.03 -0.03]
 [-0.48 -0.09 -0.16  0.11  0.   -0.  ]
 [-0.44 -0.11 -0.05  0.15  0.02 -0.21]]


In [28]:
print(pca1.explained_variance_ratio_)

[0.61 0.13 0.1  0.05 0.04 0.02]


### According to the explain variance above, the first 6 compnents account for 95% of varaince, so those are the rows that I will keep for my second clustering analysis

## Clustering after PCA analysis

In [29]:
train_kmeans2 = KMeans(n_clusters=7, max_iter=500, verbose=0, n_init=5) # initialization
train_kmeans2.fit(dTrans1)

KMeans(max_iter=500, n_clusters=7, n_init=5)

In [30]:
cluster = train_kmeans2.predict(dTrans1)

In [31]:
cluster

array([3, 3, 3, ..., 5, 5, 0], dtype=int32)

In [32]:
clusters = pd.DataFrame(cluster, columns=['Cluster'])
clusters

Cluster
0           3
1           3
2           3
3           3
4           3
...       ...
2095        0
2096        0
2097        5
2098        5
2099        0

[2100 rows x 1 columns]

In [33]:
def cluster_sizes(clusters):
    #clusters is an array of cluster labels for each instance in the data
    
    size = {}
    cluster_labels = np.unique(clusters)
    n_clusters = cluster_labels.shape[0]

    for c in cluster_labels:
        size[c] = len(dTrans1[clusters == c])
    return size

In [34]:
size = cluster_sizes(cluster)

for c in size.keys():
    print("Size of Cluster", c, "= ", size[c])

Size of Cluster 0 =  385
Size of Cluster 1 =  300
Size of Cluster 2 =  332
Size of Cluster 3 =  297
Size of Cluster 4 =  236
Size of Cluster 5 =  304
Size of Cluster 6 =  246


In [35]:
print(completeness_score(classNP,cluster))

0.6340028036465207


In [36]:
print(homogeneity_score(classNP,cluster))

0.6300797407470409


### The completeness and homogeneity score performed slightly better after PCA was performed on the dataset. The completeness score after PCA increased from 0.612 to 0.634. This shows that more rows of the same class were grouped into clusters together. The homogeneity score increased slightly to 0.663, this shows that there is a slightly lower instance of rows of different classes being clustered together.